### PR TITLE
Remove lifetime

### PR DESCRIPTION
**Pull request by Txuritan**
A mentioned in #18 the added lifetime would limit use of the API as well as cause issue with PyO3 bindings. This removes it using `Vec`s and `to_vec`s.

I do plan on using this commit as the basis for my optimization experiments which may be included depending on if this is merged or not.

Benchmark ([mimalloc](https://github.com/microsoft/mimalloc)):

<table>
<tr>
	<td>
	<td>matchit
	<td>path-tree
	<td>w/o lifetime
<tr>
	<td>insert
	<td>59µs
	<td>73µs
	<td>84µs
<tr>
	<td>find
	<td>45µs
	<td>40µs
	<td>40µs
</table>

Benchmark (Windows' allocator):

<table>
<tr>
	<td>
	<td>matchit
	<td>path-tree
	<td>w/o lifetime
<tr>
	<td>insert
	<td>109µs
	<td>128µs
	<td>158µs
<tr>
	<td>find
	<td>31µs
	<td>40µs
	<td>39µs
</table>

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -135,49 +135,50 @@
...
         Self {
             id: 0,
             routes: Vec::new(),
-            node: Node::new(NodeKind::String("".as_bytes()), None),
+            node: Node::new(NodeKind::String("".as_bytes().to_vec()), None),
         }
     }
 
-    pub fn insert(&mut self, path: &'a str, value: T) -> &mut Self {
+    pub fn insert(&mut self, path: &str, value: T) -> &mut Self {
```
Comment (adriangb) [inline src/lib.rs:170]: I think these lifetimes would be fine to keep. For PyO3 we need to copy the data into Python objects to return it anyway, so lifetimes on methods are fine. The problematic lifetimes are on the PathTree. Also I imagine most use cases don’t care too much about insertion times (ideally that would happen once at app startup) so I would _heavily_ bias toward optimizing the performance of find() over all else.

File: src/lib.rs [hunk 2]
```diff
@@ -191,7 +192,7 @@ impl<'a, T> PathTree<'a, T> {
         self
     }
 
-    pub fn find<'b>(&'a self, path: &'b str) -> Option<Path<'a, 'b, T>> {
+    pub fn find<'a, 'b>(&'a self, path: &'b str) -> Option<Path<'a, 'b, T>> {
```
Comment (adriangb) [inline src/lib.rs:195]: I think the lifetime for path would be fine to keep. For PyO3 we need to copy the data into Python objects to return it anyway, so lifetimes on methods are fine. The problematic lifetimes are on the PathTree. Also I imagine most use cases don’t care too much about insertion times (ideally that would happen once at app startup) so I would heavily bias toward optimizing the performance of find() over all else.

Comment (Txuritan) [reply src/lib.rs]: Is this change better? I decided to make it an elided lifetime instead of explicit.

Comment (adriangb) [reply src/lib.rs]: I think so, I’ll try updating the wrapper and confirm. By the way, if you have any input there feel free to open a PR, I’m still learning so I’m sure you could improve it.

**Comment by fundon** [comment]
The biggest performance problem when capturing parameters. 
Currently, I haven't found a better way to update ranges.

**Comment by fundon** [comment]
matchit uses [Params](https://docs.rs/matchit/latest/matchit/struct.Params.html) to improve performance. 
We use [smallvec](https://docs.rs/smallvec/latest/smallvec/).

**Comment by Txuritan** [comment]
From my profiling I did notice that each `find` call was an (almost) perfect triangle, means that its not taking too much time in any one check. I plan on taking a closes look using finer tracing than the initial setup I had.

I'll take a look into inserting paths, one of the ideas I had was to store the inserted paths in `PathTree` and have `Node`s reference the path by index and use `Range`s for slicing. Basically how the router works but for paths themselves.

What was the issue with capturing parameters. I did notice a second parameter check not sure if thats what you're talking about though.

**Comment by fundon** [comment]
> What was the issue with capturing parameters.
Updates parameters array, https://github.com/viz-rs/path-tree/blob/main/src/node.rs#L284-L285

**Comment by Txuritan** [comment]
Ah those, I was wondering about those. My initial thoughts were to just place them into a single insert (eg: `ranges.push(start..start + n);`) using the `Range` enum directly instead of having to recreate it when creating the raw parameteres list. I did try it out during my first pass over the code but it didn't seem like it changed much in terms of performance.

My other idea was to have each 'leaf' `Node` hold all its parameters and only build the parameter list when it found a match, but it would have complicated the router ever further.

**Comment by fundon** [comment]
Using Range is very clean and simple.

> My other idea was to have each 'leaf' Node hold all its parameters and only build the parameter list when it found a match, but it would have complicated the router ever further.

Yes we already have the path pieces that seem to do.

**Comment by fundon** [comment]
If it is possible to merge, assign to me. :)

**Comment by Txuritan** [comment]
I fixed the merge conflict but I am unable to assign you to the issue as I don't have the ability to do so.

**Comment by fundon** [comment]
@Txuritan  I have invited you to role write.

**Comment by fundon** [comment]
Landed in v0.5.1. Thanks.

**Comment by adriangb** [comment]
This worked out great for the Python wrapper: https://github.com/adriangb/routrie/commit/faa4ad8bf2a10a06ed621eb1e30a5d4cceb5cc91#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759

It would be nice to figure out at some point how to lazily build the parameters instead of calling `params()` (just like you are doing in Rust) before returning the data to Python, but I suspect it would get pretty complicated again with lifetimes for probably little benefit to users.

**Pull request merged**